homeassistant/components/esphome/voice_assistant.py

```diff
@@ -177,19 +186,119 @@ def _event_callback(self, event: PipelineEvent) -> None:
+        async with async_timeout.timeout(self.audio_timeout):
+            chunk = await self.queue.get()
+
+        while chunk:
+            segmenter.process(chunk)
+            if segmenter.in_command:
+                return True
+
+            # Buffer until command starts
+            chunk_buffer.append(chunk)
+
+            async with async_timeout.timeout(self.audio_timeout):
+                chunk = await self.queue.get()
+
+        return False
+
+    async def _segment_audio(
+        self,
+        segmenter: VoiceCommandSegmenter,
+        chunk_buffer: Sequence[bytes],
+    ) -> AsyncIterable[bytes]:
+        """Yield audio chunks until voice command has finished."""
+        # Buffered chunks first
+        for buffered_chunk in chunk_buffer:
+            yield buffered_chunk
+
+        # Timeout if no audio comes in for a while.
+        # This means the caller hung up.
+        async with async_timeout.timeout(self.audio_timeout):
+            chunk = await self.queue.get()
+
+        while chunk:
+            if not segmenter.process(chunk):
+                # Voice command is finished
+                break
+
+            yield chunk
+
+            async with async_timeout.timeout(self.audio_timeout):
+                chunk = await self.queue.get()
 
     async def run_pipeline(
         self,
+        conversation_id: str | None,
         pipeline_timeout: float = 30.0,
     ) -> None:
         """Run the Voice Assistant pipeline."""
         try:
             tts_audio_output = (
                 "raw" if self.device_info.voice_assistant_version >= 2 else "mp3"
             )
+
+            if self.device_info.voice_assistant_version >= 3:
+                segmenter = VoiceCommandSegmenter()
+                chunk_buffer: deque[bytes] = deque(maxlen=100)
+                try:
+                    async with async_timeout.timeout(pipeline_timeout):
+                        speech_detected = await self._wait_for_speech(
+                            segmenter, chunk_buffer
+                        )
+                        if not speech_detected:
+                            _LOGGER.debug("No speech detected, or cancelled")
+                            self.handle_finished()
+                            return
+                except asyncio.TimeoutError:
+                    self.handle_event(
+                        VoiceAssistantEventType.VOICE_ASSISTANT_ERROR,
+                        {
+                            "code": "speech-timeout",
+                            "message": "Timed out waiting for speech",
+                        },
+                    )
+                    self.handle_finished()
+                    return
+
+                _LOGGER.debug("Starting pipeline")
+
+                async def _stream_packets() -> AsyncIterable[bytes]:
+                    try:
+                        async for chunk in self._segment_audio(segmenter, chunk_buffer):
```

Comment:
What code is raising TimeoutError here that needs to be handled. Is that coming from segment_audio ?

Reply:
Let’s add it to the doc string of `segment_audio`